cmd/k8s-operator/spdy-hijacker.go

```diff
@@ -0,0 +1,189 @@
+			if err := closeConnWithWarning(conn, msg); err != nil {
+				return nil, multierr.New(errors.New(msg), err)
+			}
+			return nil, errors.New(msg)
+		} else {
+			msg = msg + "; failure mode is 'fail open'; continuing session without recording."
+			h.log.Warnf(msg)
+			return conn, nil
+		}
+	} else {
+		// TODO (irbekrm): log which recorder
+		h.log.Info("successfully connected to a session recorder")
+		wc = rw
+	}
+	go func() {
+		err := <-errChan
+		if err == nil {
+			h.log.Info("finished uploading the recording")
+			return
+		}
+		msg := fmt.Sprintf("connection to the session recorder errorred: %v;  failure mode set to 'fail closed'; closing connection", err)
+		h.log.Error(msg)
+		if err := closeConnWithWarning(conn, msg); err != nil {
+			h.log.Error(err)
+		}
+		return
+	}()
+	cl := tstime.DefaultClock{}
+	lc := &spdyRemoteConnRecorder{
+		log:  h.log,
+		Conn: conn,
+		lw: &loggingWriter{
+			start:           cl.Now(),
+			clock:           cl,
+			failOpen:        h.failOpen,
+			sessionRecorder: wc,
+			log:             h.log,
+		},
+	}
+
+	qp := h.req.URL.Query()
+	ch := CastHeader{
+		Version:     asciicastv2,
+		Timestamp:   lc.lw.start.Unix(),
+		ExecCommand: strings.Join(qp["command"], " "),
+		SrcNode:     strings.TrimSuffix(h.who.Node.Name, "."),
+		SrcNodeID:   h.who.Node.StableID,
+		Namespace:   h.ns,
+		Pod:         h.pod,
+	}
+	if !h.who.Node.IsTagged() {
+		ch.SrcNodeUser = h.who.UserProfile.LoginName
+		ch.SrcNodeUserID = h.who.Node.User
+	} else {
+		ch.SrcNodeTags = h.who.Node.Tags
+	}
+	lc.ch = ch
+	return lc, nil
+}
+
+// CastHeader is the asciicast header to be sent to the recorder at the start of
+// the recording of a session.
+// https://docs.asciinema.org/manual/asciicast/v2/#header
+type CastHeader struct {
+	// Version is the asciinema file format version.
+	Version int `json:"version"`
+
+	// Width is the terminal width in characters.
+	Width int `json:"width"`
+
+	// Height is the terminal height in characters.
+	Height int `json:"height"`
+
+	// Timestamp is the unix timestamp of when the recording started.
+	Timestamp int64 `json:"timestamp"`
+
+	// Tailscale-specific fields: SrcNode is the full MagicDNS name of the
+	// tailnet node originating the connection, without the trailing dot.
+	SrcNode string `json:"srcNode"`
+
+	// SrcNodeID is the node ID of the tailnet node originating the connection.
+	SrcNodeID tailcfg.StableNodeID `json:"srcNodeID"`
+
+	// SrcNodeTags is the list of tags on the node originating the connection (if any).
+	SrcNodeTags []string `json:"srcNodeTags,omitempty"`
+
+	// SrcNodeUserID is the user ID of the node originating the connection (if not tagged).
+	SrcNodeUserID tailcfg.UserID `json:"srcNodeUserID,omitempty"` // if not tagged
+
+	// SrcNodeUser is the LoginName of the node originating the connection (if not tagged).
+	SrcNodeUser string `json:"srcNodeUser,omitempty"`
+
+	// Kubernetes-specific fields:
```

Comment:
Have we got any information users can check to see which cluster the pod is in? For example, maybe we could include some information about the operator itself

Reply:
(As discussed earlier), the recorder will get send the MagicDNS name of the operator instance, which will identify the cluster